src/server.rs

```diff
@@ -58,65 +54,567 @@ use protocol::{
+        &self.service.creator
+    }
+
+    /// Returns the port that this server is bound to
+    #[allow(dead_code)]
+    pub fn port(&self) -> u16 {
+        self.listener.local_addr().unwrap().port()
+    }
+
+    /// Returns a handle to the event loop that this server is bound to.
+    #[allow(dead_code)]
+    pub fn handle(&self) -> Handle {
+        self.core.handle()
+    }
+
+    /// Runs this server to completion.
+    ///
+    /// If the `shutdown` future resolves then the server will be shut down,
+    /// otherwise the server may naturally shut down if it becomes idle for too
+    /// long anyway.
+    pub fn run<F>(self, shutdown: F) -> io::Result<()>
+        where F: Future,
+    {
+        self._run(Box::new(shutdown.then(|_| Ok(()))))
+    }
+
+    fn _run<'a>(self, shutdown: Box<Future<Item = (), Error = ()> + 'a>)
+                -> io::Result<()>
+    {
+        let SccacheServer { mut core, listener, rx, service, timeout, wait } = self;
+
+        // Create our "server future" which will simply handle all incoming
+        // connections in separate tasks.
+        let handle = core.handle();
+        let server = listener.incoming().for_each(move |(socket, _addr)| {
+            SccacheProto.bind_server(&handle, socket, service.clone());
+            Ok(())
+        });
+
+        // Right now there's a whole bunch of ways to shut down this server for
+        // various purposes. These include:
+        //
+        // 1. The `shutdown` future above.
+        // 2. An RPC indicating the server should shut down
+        // 3. A period of inactivity (no requests serviced)
+        //
+        // These are all encapsulated wih the future that we're creating below.
+        // The `ShutdownOrInactive` indicates the RPC or the period of
+        // inactivity, and this is then select'd with the `shutdown` future
+        // passed to this function.
+        let handle = core.handle();
+        let shutdown_idle = ShutdownOrInactive {
+            rx: rx,
+            timeout: Timeout::new(timeout, &handle)?,
+            handle: handle.clone(),
+            timeout_dur: timeout,
+        };
+
+        let server = server.select(shutdown_idle);
+        let shutdown = shutdown.map_err(|()| {
+            io::Error::new(io::ErrorKind::Other, "shutdown signal failed")
+        });
+        let server = server.map(|_| ()).map_err(|e| e.0).select(shutdown);
+        core.run(server)
+            .map_err(|p| p.0)?;
+
+        // Once our server has shut down either due to inactivity or a manual
+        // request we still need to give a bit of time for all active
+        // connections to finish. This `wait` future will resolve once all
+        // instances of `SccacheService` have been dropped.
+        //
+        // Note that we cap the amount of time this can take, however, as we
+        // don't want to wait *too* long.
+        core.run(wait.select(Timeout::new(Duration::new(1, 0), &handle)?))
```

Comment:
One second seems a little short for allowing real-world activity to complete. If we started a compile it could easily take longer than that to finish up and cache the result. I guess this is only an issue when you force a shutdown, so alternately we could just document somewhere (in the `--stop-server` command help?) that outstanding compilations might not finish if you ask the server to shut down.